NeoML/include/NeoML/Dnn/Dnn.h

```diff
@@ -700,6 +688,54 @@ inline CArchive& operator>>( CArchive& archive, CDnn& dnn)
 
 void NEOML_API SerializeLayer( CArchive& archive, IMathEngine& mathEngine, CPtr<CBaseLayer>& layer );
 
+//------------------------------------------------------------------------------------------------------------
+
+// This class can initialize a reference dnn, that has the same configuration as the original dnn
+// and shares parameter blobs with the original dnn to save memory.
+// Useful for multi-threaded inference where each thread can operate own reference dnn independently.
+// Learning is disabled for both the original dnn and the reference dnn.
+// Creates a copy of the original dnn's random generator to use it for inference.
+class NEOML_API CDnnReferenceRegister final {
+public:
+	// NOTE: mathEngine should live longer than CDnnReferenceRegister
+	CDnnReferenceRegister( IMathEngine& mathEngine, CDnn& dnn );
+	CDnnReferenceRegister( IMathEngine& mathEngine, CArchive& archive, int seed );
+
+	~CDnnReferenceRegister();
+
+	// Thread-safe coping of originalDnn, increments the counter
+	// NOTE: Pointer allocates memory using the `new` operator => the memory must be manually deallocated.
+	CDnn* CreateReferenceDnn();
+
+private:
+	CDnnReferenceRegister( CReferenceDnnInfo* referenceDnnInfo, IMathEngine& mathEngine );
+	// Called in destructor of reference dnn to decrease the counter and allow destroy this register class
+	void destoyReferenceDnn();
+
+	// NOTE: First field should be CDnn for CDistributedInference
+	CDnn originalDnn; // The dnn to make reference dnns
+	std::atomic<int> counter{}; // Stores the number of created reference dnns
+
+	friend class CReferenceDnnInfo;
+};
+
+//---------------------------------------------------------------------------------------------------------
+
+// Internal technical class
+class CReferenceDnnInfo final {
+public:
+	CReferenceDnnInfo( CRandom rand, CDnnReferenceRegister* ptr ) : random( rand ), registerPtr( ptr ) {}
+
+	~CReferenceDnnInfo() { if( registerPtr != nullptr ) { registerPtr->destoyReferenceDnn(); } }
+
+	CRandom& Random() { return random; }
+
+private:
+	CRandom random; // Stores the dnn's own external random class inside this dnn class
+	// For reference dnn != 0, and original dnn == 0 only
+	CDnnReferenceRegister* registerPtr = nullptr; // This pointer is not owned
```

Comment:
Did I understand correctly that this pointer is nullptr iff the dnn is originalDnn? OriginalDnn is never been used directly (private member of Register) and doesn't counted in the Register? But yet CReferenceDnnInfo* referenceDnnInfo in any originalDnn is not nullptr which enables all that asserts?

If all above is correct, why do we need such a case -- nullptr-ish CReferenceDnnInfo in OriginalDnn?

Reply:
Not exactly.

The originalDnn also can be used for inference as one of existed reference dnns (For example in `CDistributedInference`, when it is known that no more reference dnns will be created). So the originalDnn also needs to reject all non-const methods. For that used this initialized field.